stock_pull_list/wizards/stock_pull_list_wizard.py

```diff
@@ -0,0 +1,330 @@
+        if self.group_by_rule:
+            fields.append("stock_rule_id")
+        return fields
+
+    def _get_procurement_group_keys(self):
+        fields = self._get_fields_for_keys()
+        if not fields:
+            return [False]
+        options_list = []
+        for f in fields:
+            # Many2many only field type supported. As more needs arise, this
+            # can be extended
+            options_list.append(self.line_ids.mapped(f).ids)
+        return list(itertools.product(*options_list))
+
+    def _prepare_proc_group_values(self):
+        # TODO: use special secuence to name procurement groups of pull lists.
+        return {}
+
+    def action_procure(self):
+        self.ensure_one()
+        lines_obj = self.env["stock.pull.list.wizard.line"]
+        errors = []
+        proc_groups = []
+        pg_obj = self.env["procurement.group"]
+        grouping_keys = self._get_procurement_group_keys()
+        fields = self._get_fields_for_keys()
+        for gk in grouping_keys:
+            domain = [("wizard_id", "=", self.id), ("needed_qty", ">", 0.0)]
+            for i, f in enumerate(fields):
+                domain.append((f, "=", gk[i]))
+            n = 0
+            lines = lines_obj.search(domain)
+            if not lines:
+                continue
+            group = pg_obj.create(self._prepare_proc_group_values())
+            proc_groups.append(group.id)
+            procurements = []
+            for line in lines.filtered(lambda l: l.selected):
+                n += 1
+                if 0 < self.max_lines < n:
+                    n = 0
+                    group = pg_obj.create(self._prepare_proc_group_values())
+                    proc_groups.append(group.id)
+
+                values = self._prepare_procurement_values(line.date, group)
+                procurements.append(
+                    pg_obj.Procurement(
+                        line.product_id,
+                        line.needed_qty,
+                        line.product_id.uom_id,
+                        line.location_id,
+                        "Pull List %s" % self.id,
+                        "Pull List %s" % self.id,
+                        self.env.user.company_id,
+                        values,
+                    )
+                )
+            # Run procurements
+            try:
+                pg_obj.run(procurements)
+            except UserError as error:
+                errors.append(error.name)
+            if errors:
+                raise UserError("\n".join(errors))
+        res = {
+            "name": _("Generated Procurement Groups"),
+            "src_model": "stock.pull.list.wizard",
+            "view_type": "form",
+            "view_mode": "tree,form",
+            "res_model": "procurement.group",
+            "type": "ir.actions.act_window",
+            "domain": str([("id", "in", proc_groups)]),
+        }
+        return res
+
+
+class PullListWizardLine(models.TransientModel):
```

Comment:
@MateuGForgeFlow Time to split models per file ?

Reply:
Thanks for the heads up, I tend to think that wizard and wizard lines can coexist in the same file. I agree with your other comments.

Reply:
Apart of personal feelings about that because I prefer having code for one model at a time when reviewing, debugging, etc.., it is part of OCA conventions https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#file-naming

Reply:
If you do a quick overview of the modules at the OCA, the vast majority including wizards with lines have them in a single file (one example in v14: [sale_product_multi_add](https://github.com/OCA/sale-workflow/blob/14.0/sale_product_multi_add/wizards/sale_import_products.py). From this, I extract that most people find it more easy to maintain and read when it comes to wizards, for other models I do split. Can it be that sometimes conventions are not written anywhere?

Anyway, let me know if this is a blocking change for you and we would be glad to change it.

Reply:
Nope, not a blocking one.

Reply:
> the vast majority including wizards with lines have them in a single file

not an argument :smiley: